src/System.Linq/tests/SelectManyTests.cs

```diff
@@ -409,31 +409,19 @@ public void DisposeAfterEnumeration(int sourceLength, int subLength)
             bool sourceDisposed = false;
             bool[] subCollectionDisposed = new bool[sourceLength];
 
-            var sourceEnumerator = new DelegateBasedEnumerator<int>
-            {
-                MoveNextWorker = () => ++sourceState <= sourceLength,
-                CurrentWorker = () => 0,
-                DisposeWorker = () => sourceDisposed = true
-            };
-
-            var source = new DelegateBasedEnumerable<int>
-            {
-                GetEnumeratorWorker = () => sourceEnumerator
-            };
-            
-            var subEnumerator = new DelegateBasedEnumerator<int>
-            {
-                // MoveNext: Return true subLength times.
-                // Dispose: Record that Dispose was called & move to the next index.
-                MoveNextWorker = () => ++subState[subIndex] <= subLength,
-                CurrentWorker = () => subState[subIndex],
-                DisposeWorker = () => subCollectionDisposed[subIndex++] = true
-            };
-
-            var subCollection = new DelegateBasedEnumerable<int>
-            {
-                GetEnumeratorWorker = () => subEnumerator
-            };
+            DelegateIterator<int> source = null;
+            source = new DelegateIterator<int>(
+                getEnumerator: () => source,
```

Comment:
Nit: if you were to change the implementation in the ctor to:
```C
_getEnumerator = getEnumerator ?? () => this;
```
then I think you could change this to:
```C#
var source = new DelegateIterator<int>(
    moveNext: ...
```
etc., and not have to do this dance with defining the variable separately.